cmd/csi-api-gen/generators/api.go

```diff
@@ -0,0 +1,432 @@
+	return "public"
+}
+
+// packages returns a list of package generators.
+func packages(context *generator.Context, arguments *args.GeneratorArgs) (pkgs generator.Packages) {
+	// find API group definitions
+	groups := findAPIGroupDefinitions(context)
+	klog.V(5).Infof("Found API groups: %v", groups)
+
+	for _, group := range groups {
+		removeGeneratedFiles(group, arguments.OutputBase)
+		pkgs = append(pkgs, packagesForGroup(group, arguments.OutputBase)...)
+	}
+	return
+}
+
+// findAPIGroupDefinitions iterates over the context's list of package paths,
+// and builds a map mapping API group paths to their definition.
+// API group definitions are either:
+// * subdirectories of client/api
+// * or packages whose doc.go file contains a comment containing relevant tag markers
+// If the latter, see the comments on tagMarker and tagName above for supported options.
+func findAPIGroupDefinitions(context *generator.Context) []*groupDefinition {
+	pkgPaths := context.Inputs
+
+	// first, re-order the inputs by lengths, so that we always process parent packages first
+	sort.Slice(pkgPaths, func(i, j int) bool {
+		return len(pkgPaths[i]) < len(pkgPaths[j])
+	})
+
+	groups := make(map[string]*groupDefinition)
+
+	for _, pkgPath := range pkgPaths {
+		klog.V(5).Infof("Considering input %s", pkgPath)
+
+		pkg := context.Universe[pkgPath]
+		if pkg == nil {
+			// If the input had no Go files, for example.
+			continue
+		}
+
+		if buildAPIGroupDefinitionFromDocComment(pkgPath, pkg, groups) {
+			// found a +csi-proxy-api-gen comment in the package's doc.go file
+			continue
+		}
+
+		if strings.HasPrefix(pkgPath, csiProxyAPIPath) {
+			// part of the canonical API definitions, under client/api
+			buildCanonicalAPIGroupDefinition(pkgPath, pkg, groups)
+			continue
+		}
+
+		// is this package a version of an API group?
+		parts := strings.Split(pkgPath, "/")
+		parentPkgPath := strings.Join(parts[:len(parts)-1], "/")
+		if definition, present := groups[parentPkgPath]; present {
+			if !apiversion.IsValidVersion(pkg.Name) {
+				klog.Fatalf("Unexpected go package %q, should be of the form \"%s/<version>\", where <version> should be a valid API group version identifier",
+					pkg.Name, parentPkgPath)
+			}
+			klog.V(5).Infof("Found version %q for API group %q", pkg.Name, parts[len(parts)-2])
+			definition.addVersion(pkg)
+		}
+	}
+
+	result := make([]*groupDefinition, len(groups))
+	i := 0
+	for _, group := range groups {
+		// sanity check: every group should have at least one version
+		if len(group.versions) == 0 {
+			klog.Fatalf("API group %q doesn't have any version", group.name)
+		}
+		result[i] = group
+		i++
+	}
+
+	return result
+}
+
+// buildAPIGroupDefinitionFromDocComment looks for a +csi-proxy-api-gen comment in the package's
+// doc.go file, and if it finds one build the corresponding API definition.
+func buildAPIGroupDefinitionFromDocComment(pkgPath string, pkg *types.Package, groups map[string]*groupDefinition) bool {
```

Comment:
the passing pkgPath is just for logging purpose in the function?

Reply:
Just for logging.